### ISSUE TITLE
Question: Avoid requiring enzyme for `toContainReact` matcher

### ISSUE THREAD

**Issue by whawker** [blainekasten/jasmine-enzyme#29]
Hi,

I've been trying to create a Karma plugin wrapper for this project, however I've hit a road block as the `toContainReact` [matcher requires Enzyme](https://github.com/blainekasten/jasmine-enzyme/blob/master/src/assertions/toContainReact.js#L9).

I've been playing around with the matcher, and found an alternative, but hacky solution to avoid the import.

Using
```js
const wrappedInstance = Object.getPrototypeOf(enzymeWrapper).wrap(reactInstance);
```
Instead of
```js
const wrappedInstance = shallow(reactInstance);
```

This uses the [wrap method from ReactWrapper](https://github.com/airbnb/enzyme/blob/master/src/ReactWrapper.js#L855), but there may be hidden complications with the `this.root`?

Is there any way we can render the error message without importing Enzyme?

**Comment by blainekasten** [comment]
> however I've hit a road block as the toContainReact matcher requires Enzyme.

Are you not able to require enzyme? This project has a hard dependency on Enzyme.. So I'm curious why jasmine-enzyme using it internally causes issues first.

**Comment by whawker** [comment]
From what I can gather, the plugins have to run in the browser context, so `require` is undefined. One solution is to also bundle Enzyme with the plugin too, but that has requirements on React and ReactDOM, so the overall plugin becomes much larger.

**Comment by blainekasten** [comment]
I think you would probably want to use some tooling to build together all the dependencies... You could use webpack or browserify. jasmine-enzyme doesn't work without enzyme. And enzyme definitely uses `require` in it's code base also. How were you planning on working around that?

You could try something like this https://github.com/karma-runner/karma-commonjs

**Comment by whawker** [comment]
I have been using webpack to attempt to bundle the plugin.

From what I can see, apart from the _tests_ for the assertions - the only assertion that explicitly requires Enzyme is `toContainReact` - and that is only used to generate an error message.

All the other assertions have an Enzyme object injected via the `expect(<EnzymeObj>)` Jasmine call - Jasmine-Enzyme then uses the methods provided by the injected object - these assertions don't explicitly require Enzyme - just that the tested object is an Enzyme ReactWrapper.

The reason I'd like to avoid the require call in `toContainReact` is that when I bundle with that require call to Enzyme, I have to include the Enzyme library too - and all it's dependencies (React and ReactDOM), resulting in a massive bundled code file. This is why I have proposed the `getPrototypeOf` workaround, which will use methods on the injected Enzyme ReactWrapper just like the other assertions do. That way I don't need to bundle Enzyme, (and React and ReactDOM), and the Karma plugin is much much smaller.

Hope that clarifies.

**Comment by blainekasten** [comment]
> From what I can see, apart from the tests for the assertions - the only assertion that explicitly requires Enzyme is toContainReact - and that is only used to generate an error message.

It's a creative idea. But it feels unsafe unfortunately. If enzyme changed their internal implementation of what it means to contain react elements, then this idea might not stand true and result in bugs. Unfortunately I'd rather just rely on the safe method and say it's up to implementors to get it to work.

I know that's not the answer you were hoping for. But I don't think making the library less stable is a good idea just to avoid the work of bundling it up all the way.

However, looking at enzyme's docs. I might have just made a mistake anyhow. https://github.com/airbnb/enzyme/blob/master/docs/api/ShallowWrapper/contains.md
looks like contains just takes react code. Not a shallow wrapped version. Let me do some TDD and get back to you on this.

**Comment by blainekasten** [comment]
Okay. So i read the code, and your OP wrong earlier when I looked at it. I see it's because of the `.html` call. Ultimately. I think that it's unrealistic that we'll completely avoid importing enzyme for any future matchers. Even if we decided to go that way, it'd be hard to not let that slip in on accident as it _just make sense_.

So, unfortunately I'm going to close this issue and leave it to user land to figure out using this library. If there is any other ideas to make it easier i'd be intersted. Or if you want to put a post together on implementing with karma, i'd gladly put it in a doc/ folder or something.

Thanks for bringing up the issue! Good luck bundling
